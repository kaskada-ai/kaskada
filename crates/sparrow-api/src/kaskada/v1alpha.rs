#![allow(clippy::large_enum_variant)] // Could box the pulsar subscription, but for now just allowing this.
#![allow(clippy::derive_partial_eq_without_eq)]
use std::str::FromStr; // stop clippy erroring on generated code from tonic (proto)

use itertools::Itertools;
use url::Url;

use crate::kaskada::v1alpha::object_store_destination::ResultPaths;
use crate::kaskada::v1alpha::operation_plan::tick_operation::TickBehavior;

tonic::include_proto!("kaskada.kaskada.v1alpha");

/// Traits for the `Slice`
mod slice_traits;

/// Implementations and traits for parts of the plan
mod plan_impl;

/// Traits for [PreparedFile]
mod prepared_file_impl;

/// Traits for [Schema]
mod schema_traits;

impl ComputeTable {
    pub fn name(&self) -> &str {
        &self.config().name
    }

    pub fn config(&self) -> &TableConfig {
        self.config.as_ref().expect("tables should have a config")
    }
}

impl ExecuteResponse {
    /// Returns an owned vec of the output paths for this execution.
    //
    /// Returns `None` if not an `ObjectStoreOutput`.
    pub fn output_paths(&self) -> Option<Vec<String>> {
        match &self.destination {
            Some(output) => match &output.destination {
                Some(destination::Destination::ObjectStore(store)) => store
                    .output_paths
                    .as_ref()
                    .map(|output_paths| output_paths.paths.clone()),
                Some(_) | None => None,
            },
            None => None,
        }
    }

    /// Returns a mutable reference to the output paths.
    ///
    /// Returns `None` if not an `ObjectStoreOutput`.
    pub fn output_paths_mut(&mut self) -> Option<&mut ResultPaths> {
        match &mut self.destination {
            Some(output) => match &mut output.destination {
                Some(destination::Destination::ObjectStore(store)) => match &mut store.output_paths
                {
                    Some(output_path) => Some(output_path),
                    None => None,
                },
                Some(_) | None => None,
            },
            None => None,
        }
    }
}

impl FeatureSet {
    pub fn new(query: &str, formulas: Vec<(&str, &str)>) -> Self {
        Self {
            formulas: formulas
                .into_iter()
                .enumerate()
                .map(|(index, (name, formula))| Formula {
                    name: name.to_owned(),
                    formula: formula.to_owned(),
                    source_location: format!("FeatureSet::new formula {index}"),
                })
                .collect(),
            query: query.to_owned(),
        }
    }
}

impl TableConfig {
    /// Creates a table config with the given parameters and a `KaskadaSource`,
    /// which represents a file-backed table.
    pub fn new_with_table_source(
        name: &str,
        uuid: &uuid::Uuid,
        time_column_name: &str,
        subsort_column_name: Option<&str>,
        group_column_name: &str,
        grouping: &str,
    ) -> Self {
        Self {
            name: name.to_owned(),
            uuid: uuid.to_string(),
            time_column_name: time_column_name.to_owned(),
            subsort_column_name: subsort_column_name.map(|s| s.to_owned()),
            group_column_name: group_column_name.to_owned(),
            grouping: grouping.to_owned(),
            source: Some(Source {
                source: Some(source::Source::Kaskada(KaskadaSource {})),
            }),
        }
    }

    pub fn new2(
        name: &str,
        time_column_name: &str,
        subsort_column_name: Option<&str>,
        group_column_name: &str,
        grouping: &str,
    ) -> Self {
        Self::new_with_table_source(
            name,
            &uuid::Uuid::new_v4(),
            time_column_name,
            subsort_column_name,
            group_column_name,
            grouping,
        )
    }

    pub fn with_table_source(self) -> Self {
        Self::new_with_table_source(
            &self.name,
            &uuid::Uuid::from_str(&self.uuid).unwrap(),
            &self.time_column_name,
            self.subsort_column_name.as_deref(),
            &self.group_column_name,
            &self.grouping,
        )
    }
}

impl Formula {
    pub fn new(name: &str, formula: &str) -> Self {
        Self {
            name: name.to_owned(),
            formula: formula.to_owned(),
            source_location: name.to_owned(),
        }
    }
}

// Display PlanHash as the upper hex encoded vector.
impl std::fmt::Display for PlanHash {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            self.hash
                .iter()
                .format_with("", |elt, f| f(&format_args!("{elt:X}")))
        )
    }
}

impl SourceData {
    pub fn try_from_local(path: &std::path::Path) -> anyhow::Result<source_data::Source> {
        let path = match path.extension().and_then(|ext| ext.to_str()) {
            Some("parquet") => {
                source_data::Source::ParquetPath(format!("file://{}", path.display()))
            }
            Some("csv") => source_data::Source::CsvPath(format!("file://{}", path.display())),
            unsupported => anyhow::bail!("Unsupported extension {:?}", unsupported),
        };
        Ok(path)
    }

    pub fn try_from_url(url: &Url, extension: Option<&str>) -> anyhow::Result<source_data::Source> {
        let path = match extension {
            Some("parquet") => source_data::Source::ParquetPath(url.to_string()),
            Some("csv") => source_data::Source::CsvPath(url.to_string()),
            unsupported => anyhow::bail!("Unsupported extension {:?}", unsupported),
        };
        Ok(path)
    }
}

impl std::fmt::Display for FenlDiagnostics {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            Itertools::intersperse(
                self.fenl_diagnostics
                    .iter()
                    .map(|d| -> &str { &d.formatted }),
                ""
            )
            .flat_map(|f| f.lines())
            .map(|s| s.trim_end())
            .format("\n")
        )
    }
}

impl std::fmt::Display for TickBehavior {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TickBehavior::Minutely => write!(f, "minutely"),
            TickBehavior::Hourly => write!(f, "hourly"),
            TickBehavior::Daily => write!(f, "daily"),
            TickBehavior::Monthly => write!(f, "monthly"),
            TickBehavior::Yearly => write!(f, "yearly"),
            TickBehavior::Finished => write!(f, "final"),
            TickBehavior::Unspecified => panic!("Unspecified tick behavior"),
        }
    }
}

impl From<uuid::Uuid> for Uuid {
    fn from(uuid: uuid::Uuid) -> Self {
        let (high, low) = uuid.as_u64_pair();
        Self { high, low }
    }
}

impl From<&uuid::Uuid> for Uuid {
    fn from(uuid: &uuid::Uuid) -> Self {
        let (high, low) = uuid.as_u64_pair();
        Self { high, low }
    }
}

impl From<Uuid> for uuid::Uuid {
    fn from(uuid: Uuid) -> Self {
        Self::from_u64_pair(uuid.high, uuid.low)
    }
}

impl From<&Uuid> for uuid::Uuid {
    fn from(uuid: &Uuid) -> Self {
        Self::from_u64_pair(uuid.high, uuid.low)
    }
}
