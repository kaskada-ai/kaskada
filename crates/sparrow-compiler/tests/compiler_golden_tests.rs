#![warn(
    rust_2018_idioms,
    nonstandard_style,
    future_incompatible,
    clippy::mod_module_files,
    clippy::print_stdout,
    clippy::print_stderr,
    clippy::undocumented_unsafe_blocks
)]

use arrow::datatypes::DataType;
use itertools::Itertools;
use sparrow_api::kaskada::v1alpha::compile_request::ExpressionKind;
use sparrow_api::kaskada::v1alpha::schema::Field;
use sparrow_api::kaskada::v1alpha::Schema;
use sparrow_api::kaskada::v1alpha::{slice_request, SliceRequest};
use sparrow_api::kaskada::v1alpha::{
    CompileRequest, CompileResponse, ComputePlan, ComputeTable, FeatureSet, Formula,
    PerEntityBehavior, TableConfig, TableMetadata,
};
use sparrow_compiler::InternalCompileOptions;
use uuid::Uuid;

#[derive(Debug)]
struct TestScript {
    tables: Vec<TestTable>,
    feature_set: FeatureSet,
}

impl TestScript {
    async fn compile(
        self,
        slice_request: Option<SliceRequest>,
    ) -> error_stack::Result<CompileResponse, sparrow_compiler::Error> {
        let tables = self
            .tables
            .into_iter()
            .map(|table| ComputeTable {
                config: Some(TableConfig::new_with_table_source(
                    &table.name,
                    &table.id,
                    "",
                    Some(""),
                    &table.group_key,
                    &table.grouping,
                )),
                metadata: Some(TableMetadata {
                    schema: Some(table.proto_schema()),
                    file_count: 13,
                }),
                file_sets: vec![],
            })
            .collect();

        sparrow_compiler::compile_proto(
            CompileRequest {
                tables,
                feature_set: Some(self.feature_set),
                slice_request,
                expression_kind: ExpressionKind::Complete as i32,
                experimental: false,
                per_entity_behavior: PerEntityBehavior::All as i32,
            },
            InternalCompileOptions::default(),
        )
        .await
    }
}

#[derive(Debug)]
struct TestTable {
    name: String,
    id: Uuid,
    group_key: String,
    grouping: String,
    fields: Vec<TestField>,
}

impl TestTable {
    fn proto_schema(&self) -> Schema {
        let fields = self
            .fields
            .iter()
            .map(|field| Field {
                name: field.name.clone(),
                data_type: Some((&field.data_type).try_into().unwrap()),
                nullable: false,
            })
            .collect();
        Schema { fields }
    }
}

#[derive(Debug)]
struct TestField {
    name: String,
    data_type: DataType,
}

impl TestField {
    fn new(name: &str, data_type: DataType) -> Self {
        Self {
            name: name.to_owned(),
            data_type,
        }
    }
}

async fn compile(test_script: TestScript, slice_request: Option<SliceRequest>) -> ComputePlan {
    let result = test_script.compile(slice_request).await.unwrap();
    if let Some(plan) = result.plan {
        plan
    } else {
        panic!(
            "Compilation failed with diagnostics:\n{}",
            result
                .fenl_diagnostics
                .unwrap_or_default()
                .fenl_diagnostics
                .into_iter()
                .map(|elt| elt.formatted)
                .format("\n")
        )
    }
}

async fn compile_err(test_script: TestScript) -> String {
    let result = test_script.compile(None).await.unwrap();
    if let Some(plan) = result.plan {
        panic!("Expected no plan with compile errors, but was {plan:?}")
    } else {
        result.fenl_diagnostics.unwrap_or_default().to_string()
    }
}

/// Compile and return the warnings, if any.
async fn compile_warnings(test_script: TestScript) -> String {
    let result = test_script.compile(None).await.unwrap();
    let diagnostics = result.fenl_diagnostics.unwrap_or_default().to_string();

    assert!(
        result.plan.is_some(),
        "Expected plan with only compile warnings:\n\n{diagnostics}",
    );

    diagnostics
}

fn formula(name: &str, formula: &str) -> Formula {
    Formula {
        name: name.to_owned(),
        formula: formula.to_owned(),
        source_location: name.to_owned(),
    }
}

// We can't create a constant for this since it creates `String`
// which isn't (yet) possible.
fn account_sent_table() -> TestTable {
    TestTable {
        name: "Sent".to_owned(),
        id: Uuid::parse_str("111DA01F9ABD4d9d80C702AF85C822A8").unwrap(),
        group_key: "sender".to_owned(),
        grouping: "account".to_owned(),
        fields: vec![
            TestField::new("sender", DataType::UInt64),
            TestField::new("amount", DataType::Float64),
            TestField::new("receiver", DataType::UInt64),
            TestField::new("store", DataType::UInt64),
        ],
    }
}

fn primitive_table() -> TestTable {
    TestTable {
        name: "Primitive".to_owned(),
        id: Uuid::parse_str("222DA01F9ABD4d9d80C702AF85C822A8").unwrap(),
        group_key: "key".to_owned(),
        grouping: "".to_owned(),
        fields: vec![
            TestField::new("key", DataType::UInt64),
            TestField::new("f64", DataType::Float64),
            TestField::new("i64", DataType::Int64),
        ],
    }
}

fn account_received_table() -> TestTable {
    TestTable {
        name: "Received".to_owned(),
        id: Uuid::parse_str("333DA01F9ABD4d9d80C702AF85C822A8").unwrap(),
        group_key: "receiver".to_owned(),
        grouping: "account".to_owned(),
        fields: vec![
            TestField::new("amount", DataType::Float64),
            TestField::new("receiver", DataType::UInt64),
        ],
    }
}

fn store_received_table() -> TestTable {
    TestTable {
        name: "StoreReceived".to_owned(),
        id: Uuid::parse_str("444DA01F9ABD4d9d80C702AF85C822A8").unwrap(),
        group_key: "store_id".to_owned(),
        grouping: "store".to_owned(),
        fields: vec![
            TestField::new("amount", DataType::Float64),
            TestField::new("store_id", DataType::UInt64),
            TestField::new("sender_id", DataType::UInt64),
        ],
    }
}

#[tokio::test]
async fn test_basic_sum() {
    insta::assert_yaml_snapshot!(
        compile(
            TestScript {
                tables: vec![account_sent_table()],
                feature_set: FeatureSet {
                    formulas: vec![
                        formula("f1", "sum(Sent.amount)"),
                        formula("f2", "Sent.amount | sum($input)"),
                    ],
                    query: "{f1, f2}".to_owned(),
                },
            },
            None
        )
        .await
    );
}

#[tokio::test]
async fn test_projection_pushdown() {
    insta::assert_yaml_snapshot!(
        compile(
            TestScript {
                tables: vec![account_sent_table()],
                feature_set: FeatureSet {
                    formulas: vec![],
                    query: "{ amount: Sent | first() | $input.amount  }".to_owned(),
                },
            },
            None
        )
        .await
    );
}

#[tokio::test]
async fn test_unexported() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: vec![formula("f1", "sum(Sent.amount)"), formula("f2", "f1 + 1")],
                query: "{f2}".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_basic_sum_when_event() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: vec![formula("sum_amount", "sum(Sent.amount)")],
                query: "{ sum_amount } | when(Sent.amount > 10)".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}
#[tokio::test]
async fn test_count_since_window() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "count_amount",
                    "count(Sent.amount, window=since(Sent.amount > 10))",
                )],
                query: "{ count_amount }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_basic_sum_when_aggregate() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: vec![formula("sum_amount", "sum(Sent.amount)")],
                query: "{ sum_amount } | when($input.sum_amount > 10)".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_lookup_same_grouping() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), account_received_table()],
            feature_set: FeatureSet {
                formulas: vec![
                    formula("sum_amount", "sum(Sent.amount)"),
                    formula(
                        "sum_received",
                        "lookup(last(Sent.receiver), sum(Received.amount))",
                    ),
                ],
                query: "{ sum_amount, sum_received }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_with_key() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "StoreReceivedBySender",
                    "StoreReceived | with_key($input.sender_id, grouping = \"account\")",
                )],
                query: "{ sum_sent: Sent.amount | sum(), sum_store_received: \
                        StoreReceivedBySender.amount | sum() }"
                    .to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_with_key_default_grouping() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "StoreReceivedBySender",
                    "StoreReceived | with_key($input.sender_id)",
                )],
                query: "{ sum_store_received: StoreReceivedBySender.amount | sum() }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_lookup_same_grouping_with_slicing() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), account_received_table()],
            feature_set: FeatureSet {
                formulas: vec![
                    formula("sum_amount", "sum(Sent.amount)"),
                    formula(
                        "sum_received",
                        "lookup(last(Sent.receiver), sum(Received.amount))",
                    ),
                ],
                query: "{ sum_amount, sum_received }".to_owned(),
            },
        },
        Some(SliceRequest {
            slice: Some(slice_request::Slice::Percent(slice_request::PercentSlice {
                percent: 0.5,
            })),
        }),
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_lookup_diff_grouping() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![
                    formula("sum_amount", "sum(Sent.amount)"),
                    formula(
                        "sum_store_received",
                        "lookup(last(Sent.store), sum(StoreReceived.amount))",
                    ),
                ],
                query: "{ sum_amount, sum_store_received }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_cast_in_record() {
    let compute_plan = compile(
        TestScript {
            tables: vec![primitive_table()],
            feature_set: FeatureSet {
                formulas: Vec::new(),
                query: "{ x: Primitive.i64 } | else ({ x: Primitive.f64 })".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_null_value() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: Vec::new(),
                query: "{ x: null | else(Sent.amount) }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_null_record_field() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: Vec::new(),
                query: "null | else({ a: Sent.amount })".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_null_record() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table()],
            feature_set: FeatureSet {
                formulas: Vec::new(),
                query: "{a: Sent.amount, b: null } | else({ a: Sent.amount, b: 7 })".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_valid_time_domains() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "shift",
                    "Sent.amount | shift_until(Sent.amount > 0) | shift_until(Sent.amount < 24)",
                )],
                query: "{ a: shift + shift }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_since_daily() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "daily",
                    "count(Sent.amount, window=since(daily()))",
                )],
                query: "{ a: daily }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_since_daily_multiple_passes() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![formula(
                    "count_since",
                    "count(Sent.amount, window=since(daily()))",
                )],
                query: "{ a: count_since | shift_until(Sent.amount > 10) | sum($input, \
                        window=since(daily())) }"
                    .to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_tab_characters_are_allowed() {
    let compute_plan = compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![],
                query: "	{ a: 	sum(Sent.amount) }".to_owned(),
            },
        },
        None,
    )
    .await;

    insta::assert_yaml_snapshot!(compute_plan);
}

#[tokio::test]
async fn test_extend_infers_implicit_input() {
    // Tests that we don't get any diagnostic warnings when implicitly
    // inferring input using an extend.
    //
    // We don't need to assert the plan yaml, since `compile` implicitly
    // asserts that no diagnostic warnings were written.
    let query = "Sent | extend({ sum_extend: sum(Sent.amount) })";
    compile(
        TestScript {
            tables: vec![account_sent_table(), store_received_table()],
            feature_set: FeatureSet {
                formulas: vec![],
                query: query.to_owned(),
            },
        },
        None,
    )
    .await;
}

#[tokio::test]
async fn test_incompatible_err() {
    let error = compile_err(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "Sent.amount == \"hello\"".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0015]: Incompatible argument types
      --> Query:1:13
      |
    1 | Sent.amount == "hello"
      | ----------- ^^ ------- Type: string
      | |           |
      | |           Incompatible types for call to 'eq'
      | Type: f64
    "###);
}

#[tokio::test]
async fn test_incompatible_shift_functions() {
    let warnings = compile_warnings(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{x: (Sent.amount | shift_until($input > 0)) + (Sent.amount | \
                    shift_until($input > 1)) }"
                .to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(warnings, @r###"
    warning[W2000]: Incompatible time domains
      --> Query:1:45
      |
    1 | {x: (Sent.amount | shift_until($input > 0)) + (Sent.amount | shift_until($input > 1)) }
      |     --------------------------------------- ^ --------------------------------------- Time Domain: ShiftUntil { predicate: 28 }
      |     |                                       |
      |     |                                       Incompatible time domains for operation
      |     Time Domain: ShiftUntil { predicate: 16 }
    "###);
}

#[tokio::test]
async fn test_unused_lhs_pipe() {
    let warnings = compile_warnings(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ x: Sent.amount | sum(Sent.amount) }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(warnings, @r###"
    warning[W2001]: Unused binding
      --> Query:1:18
      |
    1 | { x: Sent.amount | sum(Sent.amount) }
      |                  ^ Left-hand side of pipe not used
    "###);
}

#[tokio::test]
async fn test_unused_let_binding() {
    let warnings = compile_warnings(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "let unused = sum(Sent.amount) in { x: Sent.amount }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(warnings, @r###"
    warning[W2001]: Unused binding
      --> Query:1:5
      |
    1 | let unused = sum(Sent.amount) in { x: Sent.amount }
      |     ^^^^^^ Unused binding 'unused'
    "###);
}

#[tokio::test]
async fn test_incompatible_time_domain_tables_add() {
    let warnings = compile_warnings(TestScript {
        tables: vec![account_sent_table(), account_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{x: Sent.amount + Received.amount }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(warnings, @r###"
    warning[W2000]: Incompatible time domains
      --> Query:1:17
      |
    1 | {x: Sent.amount + Received.amount }
      |     ----------- ^ --------------- Time Domain: Table 'Received'
      |     |           |
      |     |           Incompatible time domains for operation
      |     Time Domain: Table 'Sent'
    "###);
}

#[tokio::test]
async fn test_undefined_field_err() {
    let error = compile_err(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "Sent.amt == 5".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0001]: Illegal field reference
      --> Query:1:6
      |
    1 | Sent.amt == 5
      |      ^^^ No field named 'amt'
      |
      = Nearest fields: 'amount', 'store', 'sender', 'receiver'
    "###);
}

#[tokio::test]
async fn test_invalid_field_ref_base_err() {
    let error = compile_err(TestScript {
        tables: vec![account_sent_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "Sent.amount.foo".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0001]: Illegal field reference
      --> Query:1:1
      |
    1 | Sent.amount.foo
      | ^^^^^^^^^^^ No fields for non-record base type f64
    "###);
}

#[tokio::test]
async fn test_incompatible_grouping_add() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ amount: sum(Received.amount) + sum(StoreReceived.amount) }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[W2001]: Incompatible grouping
      --> Query:1:32
      |
    1 | { amount: sum(Received.amount) + sum(StoreReceived.amount) }
      |           -------------------- ^ ------------------------- Grouping: 'store'
      |           |                    |
      |           |                    Incompatible grouping for operation 'add'
      |           Grouping: 'account'
    "###);
}

#[tokio::test]
async fn test_incompatible_grouping_record() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ sum_account: sum(Received.amount), sum_store: sum(StoreReceived.amount) }"
                .to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[W2001]: Incompatible grouping
      --> Query:1:1
      |
    1 | { sum_account: sum(Received.amount), sum_store: sum(StoreReceived.amount) }
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      | |              |                                |
      | |              |                                Grouping: 'store'
      | |              Grouping: 'account'
      | Incompatible grouping for operation 'record creation'
    "###);
}

#[tokio::test]
async fn test_duplicate_formula() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![
                formula("a", "Received.amount"),
                formula("a", "Received.amount"),
            ],
            query: "{ a }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0004]: Formula already defined
     = Formula name 'a' is defined at index 0 and again at 1.
    "###);
}

#[tokio::test]
async fn test_undefined_formula() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ a }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0006]: Unbound reference
      --> Query:1:3
      |
    1 | { a }
      |   ^ No reference named 'a'
      |
      = Nearest matches: 'Received', 'StoreReceived'
    "###);
}

// Tests that a function that is rewritten into another fenl function
// correctly displays the error on the original signature.
#[tokio::test]
async fn test_invalid_rewritten_function() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "{ a: sqrt(Sent.amount, Sent.amount) }".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0008]: Invalid arguments
      --> Query:1:24
      |
    1 | { a: sqrt(Sent.amount, Sent.amount) }
      |                        ^^^^^^^^^^^ First unexpected argument
      |
      = Expected 1 but got 2
    "###);
}

#[tokio::test]
async fn test_output_type_is_struct_diagnostic() {
    let error = compile_err(TestScript {
        tables: vec![account_received_table(), store_received_table()],
        feature_set: FeatureSet {
            formulas: vec![],
            query: "Received.amount".to_owned(),
        },
    })
    .await;

    insta::assert_snapshot!(error, @r###"
    error[E0013]: Invalid output type
     = Output type must be a record, but was f64
    "###);
}
