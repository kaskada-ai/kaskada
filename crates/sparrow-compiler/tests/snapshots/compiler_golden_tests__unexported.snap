---
source: crates/sparrow-compiler/tests/compiler_golden_tests.rs
expression: compute_plan
---
per_entity_behavior: 1
operations:
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: sender
                  data_type:
                    kind:
                      Primitive: 10
                  nullable: true
                - name: amount
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
                - name: receiver
                  data_type:
                    kind:
                      Primitive: 10
                  nullable: true
                - name: store
                  data_type:
                    kind:
                      Primitive: 10
                  nullable: true
        output: false
        operator:
          Input:
            producing_operation: 0
            input_column: 0
            interpolation: 1
            column:
              ScanRecord: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 24
        output: false
        operator:
          LateBound: 1
      - arguments: []
        result_type:
          kind:
            Primitive: 13
        output: false
        operator:
          Literal:
            literal:
              Float64: 1
      - arguments: []
        result_type:
          kind:
            Primitive: 1
        output: false
        operator:
          Literal:
            literal: ~
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: amount
      - arguments:
          - 0
          - 4
        result_type:
          kind:
            Primitive: 13
        output: false
        operator:
          Instruction: field_ref
      - arguments:
          - 5
          - 3
          - 3
        result_type:
          kind:
            Primitive: 13
        output: false
        operator:
          Instruction: sum
      - arguments:
          - 6
          - 2
        result_type:
          kind:
            Primitive: 13
        output: false
        operator:
          Instruction: add
      - arguments: []
        result_type:
          kind:
            Primitive: 14
        output: false
        operator:
          Literal:
            literal:
              Utf8: f2
      - arguments:
          - 8
          - 7
        result_type:
          kind:
            Struct:
              fields:
                - name: f2
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
        output: true
        operator:
          Instruction: record
      - arguments:
          - 9
        result_type:
          kind:
            Primitive: 24
        output: false
        operator:
          Instruction: time_of
      - arguments:
          - 10
          - 1
        result_type:
          kind:
            Primitive: 2
        output: true
        operator:
          Instruction: gte
    operator:
      Scan:
        table_id:
          high: 1233317930569911709
          low: 9279388510107214504
        schema:
          fields:
            - name: sender
              data_type:
                kind:
                  Primitive: 10
              nullable: true
            - name: amount
              data_type:
                kind:
                  Primitive: 13
              nullable: true
            - name: receiver
              data_type:
                kind:
                  Primitive: 10
              nullable: true
            - name: store
              data_type:
                kind:
                  Primitive: 10
              nullable: true
        slice_plan:
          table_name: Sent
          slice: ~
  - expressions:
      - arguments: []
        result_type:
          kind:
            Struct:
              fields:
                - name: f2
                  data_type:
                    kind:
                      Primitive: 13
                  nullable: true
        output: true
        operator:
          Input:
            producing_operation: 0
            input_column: 3
            interpolation: 1
            column:
              ProducerExpression: 9
    operator:
      Select:
        input: 0
        condition:
          producing_operation: 0
          input_column: 4
          interpolation: 1
          column:
            ProducerExpression: 11
primary_grouping: account
primary_grouping_key_type:
  kind:
    Primitive: 10

